zict/lru.py

```diff
@@ -124,34 +152,42 @@ def evict(self) -> tuple[KT, VT, float]:
         -------
         Tuple of (key, value, weight)
         """
-        try:
-            key = next(iter(self.order))
-        except StopIteration:
-            raise KeyError("evict(): dictionary is empty")
-        value = self.d.pop(key)
-
-        try:
-            for cb in self.on_evict:
-                cb(key, value)
-        except Exception:
-            # e.g. if a callback tried storing to disk and raised a disk full error
-            self.d[key] = value
-            raise
-
-        del self.order[key]
+        orig_key = key
+        if key is nodefault:
+            key = self._next_eviction()
+
+        # For the purpose of multithreaded access, it's important that the value remains
+        # in self.d until all callbacks are successful.
+        # When this is used inside a Buffer, there must never be a moment when the key
+        # is neither in fast nor in slow.
+        value = self.d.get(key, nodefault)  # type: ignore[arg-type]
+        if orig_key is nodefault and value is nodefault:
+            # Race condition caused by multithreading
+            raise KeyError("evict(): dictionary is empty")  # pragma: nocover
+        elif value is nodefault:
+            raise KeyError(key)
+
+        # If we are evicting a heavy key we just inserted and one of the callbacks
+        # fails, put it at the bottom of the LRU instead of the top. This way lighter
+        # keys will have a chance to be evicted first and make space.
+        self.heavy.discard(key)
```

Comment:
Again, very opinable behaviour that is here chiefly to retain functional compatibility with main.